File: sdk/src/main/java/com/libreshockwave/cast/MemberType.java
package com.libreshockwave.cast;

/**
 * Types of cast members in Director.
 */
public enum MemberType {
    NULL(0, "null"),
    BITMAP(1, "bitmap"),
    FILM_LOOP(2, "filmLoop"),
    TEXT(3, "text"),
    PALETTE(4, "palette"),
    PICTURE(5, "picture"),
    SOUND(6, "sound"),
    BUTTON(7, "button"),
    SHAPE(8, "shape"),
    MOVIE(9, "movie"),
    DIGITAL_VIDEO(10, "digitalVideo"),
    SCRIPT(11, "script"),
    RICH_TEXT(12, "richText"),
    TRANSITION(14, "transition"),
    XTRA(15, "xtra"),
    FONT(16, "font"),
    SHOCKWAVE_3D(17, "shockwave3d"),
    UNKNOWN(255, "unknown");

    private final int code;
    private final String name;

    MemberType(int code, String name) {
        this.code = code;
        this.name = name;
    }

    public int getCode() {
        return code;
    }

    public String getName() {
        return name;
    }

    public static MemberType fromCode(int code) {
        for (MemberType type : values()) {
            if (type.code == code) {
                return type;
            }
        }
        return UNKNOWN;
    }

    @Override
    public String toString() {
        return name;
    }
}

File: cpp/src/cast/MemberType.cpp
#include "libreshockwave/cast/MemberType.hpp"

namespace libreshockwave::cast {

int code(MemberType type) {
    return static_cast<int>(type);
}

std::string_view name(MemberType type) {
    switch (type) {
        case MemberType::Null: return "null";
        case MemberType::Bitmap: return "bitmap";
        case MemberType::FilmLoop: return "filmLoop";
        case MemberType::Text: return "text";
        case MemberType::Palette: return "palette";
        case MemberType::Picture: return "picture";
        case MemberType::Sound: return "sound";
        case MemberType::Button: return "button";
        case MemberType::Shape: return "shape";
        case MemberType::Movie: return "movie";
        case MemberType::DigitalVideo: return "digitalVideo";
        case MemberType::Script: return "script";
        case MemberType::RichText: return "richText";
        case MemberType::Transition: return "transition";
        case MemberType::Xtra: return "xtra";
        case MemberType::Font: return "font";
        case MemberType::Shockwave3D: return "shockwave3d";
        case MemberType::Unknown: return "unknown";
    }
    return "unknown";
}

MemberType memberTypeFromCode(int value) {
    switch (value) {
        case 0: return MemberType::Null;
        case 1: return MemberType::Bitmap;
        case 2: return MemberType::FilmLoop;
        case 3: return MemberType::Text;
        case 4: return MemberType::Palette;
        case 5: return MemberType::Picture;
        case 6: return MemberType::Sound;
        case 7: return MemberType::Button;
        case 8: return MemberType::Shape;
        case 9: return MemberType::Movie;
        case 10: return MemberType::DigitalVideo;
        case 11: return MemberType::Script;
        case 12: return MemberType::RichText;
        case 14: return MemberType::Transition;
        case 15: return MemberType::Xtra;
        case 16: return MemberType::Font;
        case 17: return MemberType::Shockwave3D;
        default: return MemberType::Unknown;
    }
}

} // namespace libreshockwave::cast
